core/src/Dscp/Core/Transactions/Types.hs

```diff
@@ -0,0 +1,276 @@
+-- key. The second element is there to authenticate proposed changes
+-- (@kirill.andreev). Public key hash should be equal to the input address.
+-- Also, public key should be the same which used to validate signature.
+data TxWitness = TxWitness
+    { txwSig :: Signature (TxId, PublicKey, ())
+    , txwPk  :: PublicKey
+    } deriving (Eq, Show, Generic)
+
+instance Buildable TxWitness where
+    build TxWitness {..} =
+        "TxWitness { " +| txwSig |+ ", pk: " +| txwPk |+ " }"
+
+-- | Incoming publication message; transaction payload.
+data Publication blockHash addr = Publication
+    { pPrivateBlockHash  :: blockHash
+    , pPreviousBlockHash :: Maybe blockHash
+    }
+    deriving (Eq, Ord, Show, Generic)
+
+data PublicationTxWitness = PublicationTxWitness
+    { pwSig :: Signature (PublicationTxId, PublicKey, Publication (Hash PrivateBlock) Address)
+    , pwPk  :: PublicKey
+    } deriving (Eq, Show, Generic)
+
+instance Buildable PublicationTxWitness where
+    build PublicationTxWitness {..} =
+        "PublicationTxWitness { " +| pwSig |+ ", pk: " +| pwPk |+ " }"
+
+-- | Transaction coupled with witness.
+data TxWitnessed = TxWitnessed
+    { twTx      :: Tx
+    , twWitness :: TxWitness
+    } deriving (Eq, Show, Generic)
+
+instance Buildable TxWitnessed where
+    build TxWitnessed {..} =
+        "TxWitnessed { " +| twTx |+ ", " +| twWitness |+  " }"
+
+-- | Transaction coupled with witness.
+data PublicationTxWitnessed = PublicationTxWitnessed
+    { ptwTx      :: PublicationTx
+    , ptwWitness :: PublicationTxWitness
+    } deriving (Eq, Show, Generic)
+
+instance Buildable PublicationTxWitnessed where
+    build PublicationTxWitnessed {..} =
+        "PublicationTxWitnessed { " +| ptwTx |+ ", " +| ptwWitness |+  " }"
+
+-- | Actual structure in the storage.
+-- |
+-- | For each educator, we'll have a `PublicationOf id ~> LastPublication blockHash`
+-- | (which is effectively `(educator, lastBlockHash)`)
+-- | to determine the lst publication.
+newtype PublicationsOf addr
+    = PublicationsOf addr
+    deriving (Eq, Ord, Show, Generic)
+
+instance Buildable a => Buildable (PublicationsOf a) where
+    build (PublicationsOf addr) =
+        "PublicationsOf { " +| addr |+  " }"
+
+-- | Points to the `PublicationHead addr` in the database.
+data LastPublication blockHash
+    = LastPublication blockHash
+    deriving (Eq, Ord, Show, Generic)
+
+-- | Once 'LastPublication' is known, you can walk the chain of
+-- | `PublicationHead bh ~> PublicationNext bh`,
+-- | (which is `(blockHash, Maybe blockHash)`)
+-- | where phead contains block hash and pnext has prev block hash.
+newtype PublicationHead blockHash
+    = PublicationHead blockHash
+    deriving (Eq, Ord, Show, Generic)
+
+instance Buildable a => Buildable (PublicationHead a) where
+    build (PublicationHead blk) =
+        "PublicationHead { " +| blk |+  " }"
+
+data PublicationNext blockHash
+    = PublicationNext (Maybe blockHash)
```

Comment:
Please add a doc comment here (meaning of `Maybe blockHash` is not clear to me even with comment for `Publication Head` :pensive:)